File: frontend/src/lib/generateAttestationPdf.js
import { jsPDF } from "jspdf";

import appIcon from "../assets/app-icon.png";
import { abbreviateEtablissement, formatEmissionDate } from "./attestationFormat";

// ── Couleurs design ───────────────────────────────────────────────────────────
const C = {
  darkGreen:  "#1B4D3E",
  teal:       "#2E7D6B",
  bgGrey:     "#F4F7F6",
  border:     "#D1DDD9",
  textDark:   "#1A2E28",
  textMid:    "#5A7A70",
  lightGreen: "#90C4B8",
};

// ── Helpers ───────────────────────────────────────────────────────────────────
function hexToRgb(hex) {
  const h = hex.replace("#", "");
  return {
    r: parseInt(h.slice(0, 2), 16),
    g: parseInt(h.slice(2, 4), 16),
    b: parseInt(h.slice(4, 6), 16),
  };
}

function setFill(doc, hex) {
  const { r, g, b } = hexToRgb(hex);
  doc.setFillColor(r, g, b);
}

function setText(doc, hex) {
  const { r, g, b } = hexToRgb(hex);
  doc.setTextColor(r, g, b);
}

function setDraw(doc, hex) {
  const { r, g, b } = hexToRgb(hex);
  doc.setDrawColor(r, g, b);
}

function pdfSafeText(value) {
  if (value == null || value === "") return "—";
  return String(value)
    .normalize("NFC")
    .replace(/[\u00a0\u202f\u2007\u2060]/g, " ");
}

function formatMruAmount(amount) {
  const n = Math.max(0, Math.round(Number(amount) || 0));
  // Séparateur de milliers sans espace insécable (jsPDF ne le supporte pas)
  return n.toString().replace(/\B(?=(\d{3})+(?!\d))/g, " ") + " MRU";
}

function loadImageDataUrl(src) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => {
      const canvas = document.createElement("canvas");
      canvas.width = img.naturalWidth || 48;
      canvas.height = img.naturalHeight || 48;
      const ctx = canvas.getContext("2d");
      if (!ctx) { reject(new Error("Canvas indisponible")); return; }
      ctx.drawImage(img, 0, 0);
      resolve(canvas.toDataURL("image/png"));
    };
    img.onerror = () => reject(new Error("Image introuvable"));
    img.src = src;
  });
}

async function loadImageSafely(src) {
  try { return await loadImageDataUrl(src); }
  catch { return null; }
}

// ── Composants PDF ────────────────────────────────────────────────────────────

/** Cellule de la grille étudiant */
function drawCell(doc, x, y, w, h, label, value) {
  setText(doc, C.textMid);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(7);
  doc.text(label.toUpperCase(), x + 6, y + 7);

  setText(doc, C.textDark);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(10);
  const lines = doc.splitTextToSize(pdfSafeText(value), w - 10);
  doc.text(lines, x + 6, y + 13);
}

/** Petit rectangle gris clair (date / référence) */
function drawMetaBox(doc, x, y, w, h, label, value) {
  setFill(doc, C.bgGrey);
  setDraw(doc, C.border);
  doc.setLineWidth(0.25);
  doc.roundedRect(x, y, w, h, 2, 2, "FD");

  setText(doc, C.textMid);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(7);
  doc.text(label.toUpperCase(), x + 5, y + 6);

  setText(doc, C.textDark);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(10);
  doc.text(pdfSafeText(value), x + 5, y + 13);
}

// ── Export principal ──────────────────────────────────────────────────────────
export async function generateAttestationPdf(data) {
  const logoData = await loadImageSafely(appIcon);

  const doc   = new jsPDF({ unit: "mm", format: "a4" });
  const pageW = doc.internal.pageSize.getWidth();
  const pageH = doc.internal.pageSize.getHeight();
  const margin = 14;
  const contentW = pageW - margin * 2;
  let y = 0;

  // Fond blanc
  doc.setFillColor(255, 255, 255);
  doc.rect(0, 0, pageW, pageH, "F");

  // ── 1. EN-TÊTE ────────────────────────────────────────────────────────────
  const headerH = 22;
  setFill(doc, C.darkGreen);
  doc.rect(0, 0, pageW, headerH, "F");

  if (logoData) {
    doc.addImage(logoData, "PNG", margin, 3, 16, 16);
  }

  doc.setTextColor(255, 255, 255);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(16);
  doc.text("SEHILY", margin + 20, 11);

  setText(doc, C.lightGreen);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(8);
  doc.text("Bourses universitaires, simplifiées", margin + 20, 18);

  y = headerH + 10;

  // ── 2. TITRE ──────────────────────────────────────────────────────────────
  setText(doc, C.textDark);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(18);
  doc.text("ATTESTATION DE BOURSE", pageW / 2, y, { align: "center" });
  y += 5;

  // Ligne fine sous le titre
  setDraw(doc, C.teal);
  doc.setLineWidth(0.5);
  doc.line(margin + 15, y, pageW - margin - 15, y);
  y += 5;

  // Sous-titre CNOU
  setText(doc, C.textMid);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(8);
  doc.text("Centre National des Œuvres Universitaires — CNOU", pageW / 2, y, { align: "center" });
  y += 8;

  // Texte d'introduction
  doc.setFontSize(9.5);
  setText(doc, "#444444");
  doc.text(
    "Nous attestons que l'étudiant(e) ci-dessous bénéficie d'une bourse universitaire",
    pageW / 2, y, { align: "center" },
  );
  y += 5;
  doc.text(
    "pour l'année académique indiquée.",
    pageW / 2, y, { align: "center" },
  );
  y += 12;

  // ── 3. BLOC INFORMATIONS ÉTUDIANT ─────────────────────────────────────────
  const rowCount = 3;
  const gridH   = 21 * rowCount; // 21 mm par ligne
  const rowH    = gridH / rowCount;
  const colW    = (contentW - 1) / 2; // -1 pour le séparateur vertical

  setFill(doc, C.bgGrey);
  setDraw(doc, C.border);
  doc.setLineWidth(0.3);
  doc.roundedRect(margin, y, contentW, gridH, 3, 3, "FD");

  // Séparateurs horizontaux
  doc.setLineWidth(0.22);
  setDraw(doc, C.border);
  for (let i = 1; i < rowCount; i++) {
    doc.line(margin + 5, y + rowH * i, margin + contentW - 5, y + rowH * i);
  }

  // Séparateur vertical central
  doc.line(margin + colW, y + 4, margin + colW, y + gridH - 4);

  const etabAbbr = abbreviateEtablissement(data.etablissement);
  const col1x = margin;
  const col2x = margin + colW + 1;

  drawCell(doc, col1x,  y,             colW, rowH, "Nom complet",         data.nomComplet);
  drawCell(doc, col2x,  y,             colW, rowH, "Établissement",        etabAbbr);
  drawCell(doc, col1x,  y + rowH,      colW, rowH, "Numéro NNI",          data.nni);
  drawCell(doc, col2x,  y + rowH,      colW, rowH, "Niveau académique",    data.niveau);
  drawCell(doc, col1x,  y + rowH * 2,  colW, rowH, "Filière",             data.filiere);
  drawCell(doc, col2x,  y + rowH * 2,  colW, rowH, "Année universitaire", data.anneeUniversitaire);

  y += gridH + 8;

  // ── 4. MONTANT DE LA BOURSE ───────────────────────────────────────────────
  const amountH = 24;
  setFill(doc, C.darkGreen);
  doc.roundedRect(margin, y, contentW, amountH, 3, 3, "F");

  setText(doc, C.lightGreen);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(7.5);
  doc.text("MONTANT MENSUEL DE LA BOURSE", margin + 8, y + 9);

  doc.setTextColor(255, 255, 255);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(22);
  doc.text(formatMruAmount(data.montantBourse), margin + 8, y + 19);

  y += amountH + 8;

  // ── 5. DATE & RÉFÉRENCE ───────────────────────────────────────────────────
  const metaH = 19;
  const metaW = (contentW - 5) / 2;
  const emissionDate = pdfSafeText(formatEmissionDate(data.dateEmission));
  const ref = pdfSafeText(data.reference || `ATT-${Date.now()}`);

  drawMetaBox(doc, margin,             y, metaW, metaH, "Date d'émission", emissionDate);
  drawMetaBox(doc, margin + metaW + 5, y, metaW, metaH, "Référence",       ref);

  y += metaH + 14;

  // ── 6. PIED DE PAGE ───────────────────────────────────────────────────────
  setDraw(doc, C.border);
  doc.setLineWidth(0.3);
  doc.line(margin, y, pageW - margin, y);
  y += 7;

  setText(doc, C.textMid);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(8);
  doc.text(
    "Centre National des Œuvres Universitaires — Nouakchott, République Islamique de Mauritanie",
    pageW / 2, y, { align: "center" },
  );
  y += 5;
  doc.text(
    "Document généré électroniquement par la plateforme Sehily.",
    pageW / 2, y, { align: "center" },
  );

  const filename = `attestation-bourse-${ref.replace(/[^a-zA-Z0-9-]/g, "")}.pdf`;
  doc.save(filename);
}

File: mobile_app/lib/features/student/domain/generate_attestation_pdf.dart
import 'dart:typed_data';

import 'package:flutter/services.dart' show rootBundle;
import 'package:intl/date_symbol_data_local.dart';
import 'package:intl/intl.dart';
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:printing/printing.dart';

import 'attestation_format.dart';

// ── Palette ───────────────────────────────────────────────────────────────────
const _darkGreen  = PdfColor.fromInt(0xFF1B4D3E);
const _teal       = PdfColor.fromInt(0xFF2E7D6B);
const _bgGrey     = PdfColor.fromInt(0xFFF4F7F6);
const _border     = PdfColor.fromInt(0xFFD1DDD9);
const _textDark   = PdfColor.fromInt(0xFF1A2E28);
const _textMid    = PdfColor.fromInt(0xFF5A7A70);
const _lightGreen = PdfColor.fromInt(0xFF90C4B8);
const _white      = PdfColor.fromInt(0xFFFFFFFF);
const _bodyText   = PdfColor.fromInt(0xFF444444);

// ── Data class ────────────────────────────────────────────────────────────────
class AttestationPdfData {
  const AttestationPdfData({
    this.nomComplet,
    this.nni,
    this.etablissement,
    this.filiere,
    this.niveau,
    this.anneeUniversitaire,
    this.montantBourse,
    this.reference,
    this.dateEmission,
  });

  final String? nomComplet;
  final String? nni;
  final String? etablissement;
  final String? filiere;
  final String? niveau;
  final String? anneeUniversitaire;
  final double? montantBourse;
  final String? reference;
  final String? dateEmission;
}

// ── Helpers ───────────────────────────────────────────────────────────────────
Future<Uint8List?> _loadAsset(String path) async {
  try {
    final d = await rootBundle.load(path);
    return d.buffer.asUint8List();
  } catch (_) {
    return null;
  }
}

String _safe(String? v) =>
    (v == null || v.trim().isEmpty) ? '\u2014' : v.trim();

String _formatDate(String? iso) {
  if (iso != null && iso.trim().isNotEmpty) {
    final dt = DateTime.tryParse(iso);
    if (dt != null) {
      return DateFormat('dd MMMM yyyy', 'fr_FR').format(dt.toLocal());
    }
  }
  return DateFormat('dd MMMM yyyy', 'fr_FR').format(DateTime.now());
}

String _formatMontant(double? amount) {
  final n = (amount ?? 0).round();
  final s = n.toString();
  final buf = StringBuffer();
  for (int i = 0; i < s.length; i++) {
    if (i > 0 && (s.length - i) % 3 == 0) buf.write(' ');
    buf.write(s[i]);
  }
  return '$buf MRU';
}

// ── Widgets PDF ───────────────────────────────────────────────────────────────

pw.Widget _labelValue(pw.Font boldFont, String label, String value) {
  return pw.Column(
    crossAxisAlignment: pw.CrossAxisAlignment.start,
    children: [
      pw.Text(
        label.toUpperCase(),
        style: pw.TextStyle(color: _textMid, fontSize: 7),
      ),
      pw.SizedBox(height: 3),
      pw.Text(
        value,
        style: pw.TextStyle(
          font: boldFont,
          color: _textDark,
          fontSize: 10,
        ),
      ),
    ],
  );
}

pw.Widget _studentGrid(pw.Font boldFont, {
  required String nom,
  required String nni,
  required String filiere,
  required String etablissement,
  required String niveau,
  required String annee,
}) {
  final divH = pw.Container(height: 0.4, color: _border);

  pw.Widget row(String l1, String v1, String l2, String v2) {
    return pw.Padding(
      padding: const pw.EdgeInsets.symmetric(vertical: 8, horizontal: 10),
      child: pw.Row(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          pw.Expanded(child: _labelValue(boldFont, l1, v1)),
          pw.Container(width: 0.4, color: _border),
          pw.SizedBox(width: 10),
          pw.Expanded(child: _labelValue(boldFont, l2, v2)),
        ],
      ),
    );
  }

  return pw.Container(
    decoration: pw.BoxDecoration(
      color: _bgGrey,
      border: pw.Border.all(color: _border, width: 0.4),
      borderRadius: pw.BorderRadius.circular(6),
    ),
    child: pw.Column(
      children: [
        row('Nom complet', nom, '\u00c9tablissement', etablissement),
        divH,
        row('Num\u00e9ro NNI', nni, 'Niveau acad\u00e9mique', niveau),
        divH,
        row('Fili\u00e8re', filiere, 'Ann\u00e9e universitaire', annee),
      ],
    ),
  );
}

pw.Widget _montantBox(pw.Font boldFont, String montant) {
  return pw.Container(
    width: double.infinity,
    padding: const pw.EdgeInsets.symmetric(horizontal: 16, vertical: 14),
    decoration: pw.BoxDecoration(
      color: _darkGreen,
      borderRadius: pw.BorderRadius.circular(6),
    ),
    child: pw.Column(
      crossAxisAlignment: pw.CrossAxisAlignment.start,
      children: [
        pw.Text(
          'MONTANT MENSUEL DE LA BOURSE',
          style: pw.TextStyle(color: _lightGreen, fontSize: 7.5),
        ),
        pw.SizedBox(height: 6),
        pw.Text(
          montant,
          style: pw.TextStyle(font: boldFont, color: _white, fontSize: 22),
        ),
      ],
    ),
  );
}

pw.Widget _metaBox(pw.Font boldFont, String label, String value) {
  return pw.Expanded(
    child: pw.Container(
      padding: const pw.EdgeInsets.all(10),
      decoration: pw.BoxDecoration(
        color: _bgGrey,
        border: pw.Border.all(color: _border, width: 0.4),
        borderRadius: pw.BorderRadius.circular(4),
      ),
      child: pw.Column(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          pw.Text(
            label.toUpperCase(),
            style: pw.TextStyle(color: _textMid, fontSize: 7),
          ),
          pw.SizedBox(height: 4),
          pw.Text(
            value,
            style: pw.TextStyle(font: boldFont, color: _textDark, fontSize: 10),
          ),
        ],
      ),
    ),
  );
}

// ── Export ────────────────────────────────────────────────────────────────────
Future<void> generateAndShareAttestationPdf(AttestationPdfData data) async {
  await initializeDateFormatting('fr_FR');

  final logoBytes = await _loadAsset('assets/images/sehily-logo.png') ??
      await _loadAsset('assets/images/app_icon.png');

  final baseFont = await PdfGoogleFonts.notoSansRegular();
  final boldFont = await PdfGoogleFonts.notoSansBold();
  final theme = pw.ThemeData.withFont(base: baseFont, bold: boldFont);

  final ref = (data.reference?.trim().isNotEmpty == true)
      ? data.reference!.trim()
      : 'ATT-${DateTime.now().millisecondsSinceEpoch}';

  final montant  = _formatMontant(data.montantBourse);
  final emission = _formatDate(data.dateEmission);
  final etabAbbr = AttestationFormat.abbreviateEtablissement(data.etablissement);
  const hPad = 36.0;

  final doc = pw.Document();
  doc.addPage(
    pw.Page(
      pageTheme: pw.PageTheme(
        pageFormat: PdfPageFormat.a4,
        margin: pw.EdgeInsets.zero,
        theme: theme,
      ),
      build: (context) => pw.Column(
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        children: [
          // ── 1. EN-TÊTE (pleine largeur) ─────────────────────────────────
          pw.Container(
            width: double.infinity,
            color: _darkGreen,
            padding: const pw.EdgeInsets.symmetric(horizontal: 28, vertical: 12),
            child: pw.Row(
              crossAxisAlignment: pw.CrossAxisAlignment.center,
              children: [
                if (logoBytes != null)
                  pw.Image(pw.MemoryImage(logoBytes), width: 40, height: 40, fit: pw.BoxFit.contain)
                else
                  pw.SizedBox(width: 40, height: 40),
                pw.SizedBox(width: 14),
                pw.Column(
                  crossAxisAlignment: pw.CrossAxisAlignment.start,
                  mainAxisAlignment: pw.MainAxisAlignment.center,
                  children: [
                    pw.Text(
                      'SEHILY',
                      style: pw.TextStyle(font: boldFont, color: _white, fontSize: 18),
                    ),
                    pw.Text(
                      'Bourses universitaires, simplifi\u00e9es',
                      style: pw.TextStyle(color: _lightGreen, fontSize: 8),
                    ),
                  ],
                ),
              ],
            ),
          ),

          pw.SizedBox(height: 22),

          // ── Contenu avec padding horizontal ──────────────────────────────
          pw.Padding(
            padding: const pw.EdgeInsets.symmetric(horizontal: hPad),
            child: pw.Column(
              crossAxisAlignment: pw.CrossAxisAlignment.center,
              children: [
                // ── 2. TITRE ────────────────────────────────────────────
                pw.Text(
                  'ATTESTATION DE BOURSE',
                  textAlign: pw.TextAlign.center,
                  style: pw.TextStyle(font: boldFont, color: _textDark, fontSize: 18),
                ),
                pw.SizedBox(height: 6),
                pw.Container(height: 0.8, color: _teal),
                pw.SizedBox(height: 6),
                pw.Text(
                  'Centre National des \u0152uvres Universitaires \u2014 CNOU',
                  textAlign: pw.TextAlign.center,
                  style: pw.TextStyle(color: _textMid, fontSize: 8),
                ),
                pw.SizedBox(height: 10),
                pw.Text(
                  'Nous attestons que l\'\u00e9tudiant(e) ci-dessous b\u00e9n\u00e9ficie d\'une bourse universitaire',
                  textAlign: pw.TextAlign.center,
                  style: pw.TextStyle(color: _bodyText, fontSize: 9.5),
                ),
                pw.SizedBox(height: 4),
                pw.Text(
                  'pour l\'ann\u00e9e acad\u00e9mique indiqu\u00e9e.',
                  textAlign: pw.TextAlign.center,
                  style: pw.TextStyle(color: _bodyText, fontSize: 9.5),
                ),
                pw.SizedBox(height: 16),

                // ── 3. GRILLE ÉTUDIANT ───────────────────────────────────
                _studentGrid(
                  boldFont,
                  nom:           _safe(data.nomComplet),
                  nni:           _safe(data.nni),
                  filiere:       _safe(data.filiere),
                  etablissement: etabAbbr,
                  niveau:        _safe(data.niveau),
                  annee:         _safe(data.anneeUniversitaire),
                ),
                pw.SizedBox(height: 14),

                // ── 4. MONTANT ───────────────────────────────────────────
                _montantBox(boldFont, montant),
                pw.SizedBox(height: 14),

                // ── 5. DATE & RÉFÉRENCE ──────────────────────────────────
                pw.Row(
                  children: [
                    _metaBox(boldFont, 'Date d\'\u00e9mission', emission),
                    pw.SizedBox(width: 10),
                    _metaBox(boldFont, 'R\u00e9f\u00e9rence', ref),
                  ],
                ),
                pw.SizedBox(height: 24),

                // ── 6. PIED DE PAGE ──────────────────────────────────────
                pw.Container(height: 0.4, color: _border),
                pw.SizedBox(height: 8),
                pw.Text(
                  'Centre National des \u0152uvres Universitaires \u2014 Nouakchott, R\u00e9publique Islamique de Mauritanie',
                  textAlign: pw.TextAlign.center,
                  style: pw.TextStyle(color: _textMid, fontSize: 8),
                ),
                pw.SizedBox(height: 4),
                pw.Text(
                  'Document g\u00e9n\u00e9r\u00e9 \u00e9lectroniquement par la plateforme Sehily.',
                  textAlign: pw.TextAlign.center,
                  style: pw.TextStyle(color: _textMid, fontSize: 8),
                ),
              ],
            ),
          ),
        ],
      ),
    ),
  );

  await Printing.layoutPdf(onLayout: (_) async => doc.save());
}
